tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py

```diff
@@ -0,0 +1,161 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Functional tests for fused batch-norm related to determinism."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import numpy as np
+
+from tensorflow.python.eager import backprop
+from tensorflow.python.framework import config
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import errors_impl
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import nn_impl
+# The following import is required to register the gradient function
+from tensorflow.python.ops.nn_grad import _FusedBatchNormV3Grad # pylint: disable=unused-import
+from tensorflow.python.platform import test
+
+
+class FusedBatchNormalizationDeterministicTest(test.TestCase):
+  """Test determinsitic functionality and exceptions for FusedBatchNorm.
+
+  Test that tf.errors.UnimplementedError is thrown, as
+  appropriate, by the GPU code-path through FusedBatchNormFreezeGrad when
+  deterministic ops are enabled. This test assumes that
+  nn_fused_batchnorm_test.py runs equivalent test cases when deterministic ops
+  are not enabled and will therefore detect erroneous exception throwing in
+  those cases.
+
+  Also test that the other code-paths, running on both CPU and GPU, operate
+  deterministically.
+  """
+
+  def _genParams(self, data_format, x_dtype, large_batch):
+    if large_batch:
+      batch_size = 5000
+      height = width = 4
+    else:
+      batch_size = 10
+      height = 5
+      width = 5000
+    channel_count = 3
+    if data_format == 'NHWC':
+      x_shape = (batch_size, height, width, channel_count)
+    else: # 'NCHW'
+      x_shape = (batch_size, channel_count, height, width)
+    # Using random_ops.random_normal would produce different values on each run
+    x = constant_op.constant(
+        np.random.normal(size=x_shape), dtype=x_dtype)
+    scale_shape = (channel_count,)
+    scale = constant_op.constant(
+        np.random.normal(size=scale_shape), dtype=dtypes.float32)
+    offset = constant_op.constant(
+        np.random.normal(size=scale_shape), dtype=dtypes.float32)
+    mean = np.random.normal(size=scale_shape)
+    variance = np.random.normal(size=scale_shape)
+    y_shape = x_shape; y_dtype = x_dtype
+    upstream_gradients = constant_op.constant(
+        np.random.normal(size=y_shape), dtype=y_dtype)
+    return x, scale, offset, mean, variance, upstream_gradients
+
+  @test_util.run_in_graph_and_eager_modes
+  def testForward(self):
+    with self.cached_session():
+      np.random.seed(123)
```

Comment:
The seed is [automatically set](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/framework/test_util.py;l=2162;drc=67c19451cedcc0774154a386da21f6c42ebbe73f) in the base class's setUp method, so setting here is unncessary.

Reply:
Wonderful. Thanks for revealing that.